### ISSUE TITLE
Run the custom function on click of insert image button.

### ISSUE THREAD

**Issue by sandeepwisecells** [fraywing/textAngular#1104]
I implemented a custom upload image button which when clicked is opens a modal box to upload a image and then insert it into the editor. it is working fine only if w upload image in the editor.
if we type some text and then try to upload the image it did nothing.
here is the snippet of code.

 `app.config(function($provide) {

    $provide.decorator('taOptions', ['taRegisterTool', '$delegate', '$modal', function(taRegisterTool, taOptions, $modal) {
        taRegisterTool('uploadImage', {
            buttontext: 'Upload Image',
            iconclass: "fa fa-image",
            action: function(deferred, restoreSelection) {
                $modal.open({
                    controller: 'UploadImageModalInstance',
                    templateUrl: '/static/app/assessment/angular/partial/upload.html'
                }).result.then(
                    function(result) {
                        restoreSelection();
                        document.execCommand('insertImage', true, result);
                        deferred.resolve();
                    },
                    function() {
                        deferred.resolve();
                    }
                );
                return false;
            }
        });
        taOptions.toolbar[1].push('uploadImage');
        return taOptions;
    }]);
})`

**Comment by pbassut** [comment]
So, you're saying that if you upload without typing anything first it works just fine. But, if you type something before trying to upload, the uploading doesn't work. Correct?

**Comment by sandeepwisecells** [comment]
yeah correct. if i type some text and then try to upload it does not work ..

**Comment by pbassut** [comment]
Right. 
Since you're implementing a third-party freamwork, we'll need a sample to reproduce the issue and see where the problem lies.

**Comment by baldo903** [comment]
also experiencing same issue. 

this is my code for the selection of image from a md dialog container

                taRegisterTool('uploadImage', {
                    buttontext: 'Select Image',
                    iconclass: "fa fa-image",
                    action: function (deferred, restoreSelection) {
                        $mdDialog.show({
                            templateUrl: 'uploadImagesCont.tmpl.html',
                            parent: angular.element(document.body),
                            //controller: textAngularUploadImage,
                            clickOutsideToClose: false
                        }).then(function (result) {
                            //restoreSelection();
                            document.execCommand('insertImage', true, $rootScope.ImageResult);
                            deferred.resolve();
                        });
                        return false;
                    }
                });

**Comment by baldo903** [comment]
hi guys i was able to find a solution on my end. 

here is my code:
i used the sel.collapseToEnd();

                taRegisterTool('uploadImage', {
                    buttontext: 'Select Image',
                    iconclass: "fa fa-image",
                    action: function (deferred, restoreSelection) {
                        var self = this;
                        var sel = rangy.getSelection();
                        $mdDialog.show({
                            templateUrl: 'uploadImagesCont.tmpl.html',
                            parent: angular.element(document.body),
                            //controller: textAngularUploadImage,
                            clickOutsideToClose: false
                        }).then(function (result) {
                            sel.collapseToEnd();
                            self.$editor().wrapSelection('insertImage', $rootScope.ImageResult);
                            deferred.resolve();
                        });
                        return false;
                    }
                });

Thanks. hope this helps

**Comment by Nebulosar** [comment]
I would mark this solution as a good workaround.

**Note that this does not allow you to select and replace something, because of the collapse..**

In my opinion, if you look at what is necessary to make this work, the problem lies deeper in the wrapSelection or calling of the action function.

For now this works
Jeej custom dialogs! Booh windows prompt :P
